### PR TITLE
Add implementation for methods from ROS-183

### PR DESCRIPTION
**Pull request by DoSOfRedRiver**
The implementation for methods from https://rchain.atlassian.net/browse/ROS-183

There are few methods replaced:
- `asCString` replaced in favour to `Show` typeclass;
- `printOn`, `printQuotedOn`, `displayOn` now depends on `Show` typeclass;
- `clone`, `cloneTo` methods are replaced in favour to new `copy` method;

This PR is little bit messy because this branch contains commit from #34 
I'll add tests a bit later.

I still have problems with implementations for following:
- Methods “getAddr” and “setAddr” methods look exactly the same like “getLex” and “setLex”, except that given values are wrapped in “FIXNUM(ADDR_TO_PRE_FIXNUM(_))” and PRE_FIXNUM_TO_ADDR(FIXVAL(_))”. I’m not sure that this difference should be reflected in Scala implementation.
- Not sure that I can provide a correct implementation for "getField” and “setField” methods. Any help would be appreciated!

### CODE REVIEW

File: roscala/src/main/scala/coop/rchain/rosette/Ob.scala [hunk 1]
```diff
@@ -111,4 +224,21 @@ object Ob {
       def updateSelf[T](value: A => A): A = value(base)
     }
   }
+
+  def printOn[A <: Ob: Show](ob: A, file: File): Unit = {
```
Comment (pyrocto) [inline roscala/src/main/scala/coop/rchain/rosette/Ob.scala]: Add a TODO to use a logging framework instead

File: roscala/src/main/scala/coop/rchain/rosette/Ob.scala [hunk 2]
```diff
@@ -10,38 +12,149 @@ case object Upcall extends LookupError
...
+    ??? //TODO
+  }
+
+  def setLex(ind: Int, level: Int, offset: Int, value: Ob): Ob = {
+    val p: Ob = nthParent(level)
+
+    actorExtensionFiltered(ind, p)
+      .filter { _ => offset < p.numberOfSlots() }
+      .map { ob =>
+        ob.slot(offset) = value
```
Comment (pyrocto) [inline roscala/src/main/scala/coop/rchain/rosette/Ob.scala]: We're using lenses to avoid mutation.  Talk to Timm about how to do this.

Comment (tymm) [reply roscala/src/main/scala/coop/rchain/rosette/Ob.scala]:
```
import Lenses._

ob.update(_ >> 'slot)(_.updated(offset, value))
```
If the data structure is not deeply nested, I think `ob.copy(...)` is fine though.

The problem here however is that `ASSIGN(p, slot(offset), val);`/`ob.slot(offset) = value` is a side effects to `this`. Therefore `setLex` needs to return `(Ob, Ob)` or we allow for mutation in `Ob` for now until we figure things out.
I would argue for the latter.

**Comment by DoSOfRedRiver** [comment]
@tymm @pyrocto Could you please review the changes?
Also, can you clarify the meaning of SLOT_NUM macro? For now, I can leave implementation, I guess.

**Comment by tymm** [comment]
@DoSOfRedRiver The `SLOT_NUM(type, field)` macro will give you the index for `field` in `_slot`. So `SLOT_NUM(Actor, extension))` will tell you where `extension` is in the `_slot` array.
In `src/Ob.cc:535` it is used to check if `p` is really a valid `Actor` with an `extension`.

**Comment by DoSOfRedRiver** [comment]
@tymm Thanks! I will replace this thing with pattern matching.

File: roscala/src/main/scala/coop/rchain/rosette/utils/Instances.scala [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package coop.rchain.rosette.utils
+
+import coop.rchain.rosette.{Ob, TblObject}
+import shapeless._
+import record._
+import syntax.singleton._
+
+import scala.collection.mutable
+
+object Instances {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/utils/Instances.scala]: Can you briefly describe what you are trying to achieve here?

File: roscala/src/main/scala/coop/rchain/rosette/utils/package.scala [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import scala.collection.mutable
+
+package object utils {
+
+  def printToFile(f: File)(op: PrintWriter => Unit) {
+    val p = new PrintWriter(f)
+    try { op(p) } finally { p.close() }
+  }
+
+  def pSlice[T](src: mutable.Seq[T], start: Int, end: => Int) =
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/utils/package.scala]: Can you briefly describe what you are trying to achieve here?

File: roscala/src/main/scala/coop/rchain/rosette/utils/Instances.scala (hunk 3 above)
Comment (DoSOfRedRiver) [reply roscala/src/main/scala/coop/rchain/rosette/utils/Instances.scala]: Here is Instance that helps Shapeless to derive an instance for a lens with a selector for `slot`.

File: roscala/src/main/scala/coop/rchain/rosette/utils/package.scala (hunk 4 above)
Comment (DoSOfRedRiver) [reply roscala/src/main/scala/coop/rchain/rosette/utils/package.scala]: This class is used to provide the C++ pointer semantics for arrays, so each mutation on source affects the result of `pSlice` function.

**Pull request merged**